src/plugin.js

```diff
@@ -0,0 +1,139 @@
+      modulePromises.push(this._loadPackagePlugin(name));
+    }
+    Promise.all(modulePromises).then((vals) => {
+      const modules = vals;
+      console.log(`Found ${modules.length} plugin modules`);
+      const pluginObjPromises = [];
+      for (const mod of modules) {
+        pluginObjPromises.push(this._castToPluginObjects(mod));
+      }
+      Promise.all(pluginObjPromises).then((vals) => {
+        for (const pluginObjs of vals) {
+          this.plugins = this.plugins.concat(pluginObjs);
+        }
+        console.log(`Loaded ${this.plugins.length} plugins`);
+      });
+    });
+  }
+
+  async _loadFilePlugin(filepath) {
+    const lp = await lpModule;
+    const resolvedModuleFile = await lp.resolvePlugin(
+      path.join(process.cwd(), filepath),
+    );
+    return await lp.loadPlugin(resolvedModuleFile);
+  }
+
+  async _loadPackagePlugin(packageName) {
+    const lp = await lpModule;
+    const resolvedPackageFile = await lp.resolvePlugin(packageName, {
+      prefix: '@finos',
+    });
+    return await lp.loadPlugin(resolvedPackageFile);
+  }
+
+  async _castToPluginObjects(pluginModule) {
+    const plugins = [];
+    // iterate over the module.exports keys
+    for (const key of Object.keys(pluginModule)) {
+      // only extract instances of the GenericPlugin base class
+      // add it to the list of loaded plugins
+      if (
+        Object.prototype.hasOwnProperty.call(pluginModule, key) &&
+        pluginModule[key] instanceof ProxyPlugin
+      ) {
+        plugins.push(pluginModule[key]);
+      }
+    }
+    return plugins;
+  }
+}
+
+/**
+ * Parent class for all Git Proxy plugins. New plugin types must inherit from
+ * ProxyPlugin to be loaded by PluginManager.
+ */
+class ProxyPlugin {}
+
+/**
+ * A plugin which executes a function when receiving a proxy request.
+ */
+class GenericPlugin extends ProxyPlugin {
+  /**
+   * Setup a new instance of GenericPlugin with a custom exec function.
+   * @param {Promise<Action>} exec A Promise that executes when a push
+   *                               is proxied. It returns an Action. 
+   * @see {@}
+   */
+  constructor(exec) {
+    super();
+    this.exec = exec;
+  }
+}
+
+const setupPluginManager = () => {
+  // Auto-register plugins that are part of git-proxy core
+  let packagePlugins = [];
+  let localPlugins = ['./packages/git-proxy-notify-hello/index.js'];
+  // How should additional plugins be registered? Environment variable?
+  if (process.env.GITPROXY_LOAD_PLUGINS) {
+    const plugins = process.env.GITPROXY_LOAD_PLUGINS.split(',');
+    plugins
+      .filter((p) => {
+        p.length !== 0;
+      })
+      .forEach((p) => {
+        // need to differentiate between plugins loaded via file path and
+        // ones loaded using Node's builtin name resolution (node_modules)
+        if (p.charAt(0) !== '@' && p.includes('/')) {
+          localPlugins = localPlugins.concat(p);
+        } else {
+          packagePlugins = packagePlugins.concat(p);
```

Comment:
TODO: simply push a new item into the plugin list instead of concat. 

```suggestion
          localPlugins.push(p);
        } else {
          packagePlugins.push(p);
```